### DISCUSSION TITLE
RuntimeError: this event loop is already running

### DISCUSSION THREAD

**Discussion by foarsitter** [emmett-framework/granian#194] [General]
Got enthusiastic about this yet another rust replacement (`yarr`) when I heard about it. I like where the Python landscape is  heading to. 

Replaced `gunicorn` with `granian` and everything went smooth, so great job! 

Some third party library forced me in using asyncio since they stopped supporting the sync version. Using `asyncio.get_event_loop()` and `loop.run_until_complete` let me run there async function in my wsgi application. 

After adopting `granian` I experienced the following error. 

```
RuntimeError: this event loop is already running.
```

I can reproduced the error with the following flask snippet. localhost:8000/loop errors, while localhost:8000/lasyncio and localhost:8000/asgiref are succeeding where the later is the best approach I think.  

```python
import asyncio

from asgiref.sync import async_to_sync
from flask import Flask

app = Flask(__name__)

loop = asyncio.get_event_loop()


async def async_function():
    return "<p>Hello, World!</p>"


@app.route("/loop")
def run_loop():
    return loop.run_until_complete(async_function())


@app.route("/asyncio")
def run_asyncio():
    return asyncio.run(async_function())


@app.route("/asgiref")
def run_async_to_sync():
    return async_to_sync(async_function)()
```

My question is if I just was lucky that my `run_until_complete` version worked with gunicorn or that it is something that should work out of the box with granian?

**Comment by gi0baro** [comment]
@foarsitter the fact you got it working on gunicorn is just because it doesn't rely on asyncio event loop. Granian on the other hand uses asyncio even in WSGI protocol (just to preserve the same instances tbh) and thus the event loop is already running. `run_until_complete` expects to take full control over the event loop.

Just a small OT consideration: running sync code through asyncio is quite inefficient: I personally would consider switching to async flow and ASGI completely if feasible.

**Reply by foarsitter** [reply]
Thanks for your quick answer! Is there any difference in performance between running sync code in asgi or running async code in wsgi? Rewriting the hole will be to costly but replacing wsgi with asgi by leveraging WSGIMiddleware is feasible.
